aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java

```diff
@@ -749,4 +795,23 @@ public <T extends S3ClientBuilder> void applyEndpointConfigurations(T builder) {
       builder.endpointOverride(URI.create(endpoint));
     }
   }
+
+  /**
+   * Add the S3 Access Grants Plugin for an S3 client.
+   *
+   * <p>Sample usage:
+   *
+   * <pre>
+   *     S3Client.builder().applyMutation(s3FileIOProperties::applyS3AccessGrantsConfigurations)
+   * </pre>
+   */
+  public <T extends S3ClientBuilder> void applyS3AccessGrantsConfigurations(T builder) {
+    if (isS3AccessGrantsEnabled) {
+      S3AccessGrantsPlugin s3AccessGrantsPlugin =
```

Comment:
I think there is an issue we need to handle here. This will force the users to have a dependency of the S3AccessGrantsPlugin library, even if they do not need this feature. To circumvent this, you can take a look at how we do that for things like HTTP client library dependencies: https://github.com/apache/iceberg/blob/main/aws/src/main/java/org/apache/iceberg/aws/ApacheHttpClientConfigurations.java. By moving all dependencies of a feature to a different class, it can avoid a compile time dependency.

Reply:
While looking into this, I don't think even the ApacheHttpClientConfigurations reflection workaround that was implemented earlier truly avoids a compile time dependency to be fair. I agree it may allow us to not need to load both conflicting JARs in at runtime, but while compiling the iceberg-aws module, you'll still need both ApacheHttpClient and UrlConnectionHttpClient classes.

As a result, I'm interpreting what you're saying as: we should allow users to choose to add this dependency or not at runtime itself. But I don't think the same issues we were trying to solve in the ApacheHttpClient/UrlConnectionHttpClient case are (or ever will be) applicable here. Any number of AWS SDK Plugins can be attached on an AWS Client - and any collision in behavior between them are solely the responsibility of the customer defining the right set of plugins to add to the client (in other words, the mere presence of multiple plugins in the classpath cannot cause issues - only the runtime adding of the plugins to the client, which we cannot control anyways).

I think this leaves one last concern (in my opinion): increased file size of the iceberg-aws module JAR. When I try to exclude the unnecessary dependencies from the S3 Access Grants plugin jar, it only increases the iceberg-aws module JAR's size from ~29MB to ~31MB (an increase of ~2MB). That change is not reflected in the PR, but if we are okay with including the plugin within the iceberg-aws module jar, I can make that change to the PR quickly.

So, I think the question really boils down to: is the 2MB significant enough that it is problematic to package out of the box with Iceberg? If you think so, I can invest the time to add the reflection logic in. If not, I don't really see a reason to make any change on here.

Reply:
> but while compiling the iceberg-aws module, you'll still need both ApacheHttpClient and UrlConnectionHttpClient classes.

We are talking about runtime dependency on a package, not compile time. Yes compile time dependency is required. 2MB is still significant in my mind since we are really trying to minimize bundle size as much as possible. But the more important issue is that, people who use the current AWS bundle would get a failure unless they update their stack to include the plugin they do not use, which is going to be a blocker for upgrades.

Reply:
Since you've given the datapoint that the 2MB will not be acceptable, I'll work further on this as you've suggested previously. But I'm still curious how on your other point: if the S3 Access Grants plugin classes are bundled with the Iceberg AWS module JAR, then how does this make a failure for any user at all? The required classes will always be present in the classpath for every user then. 

Unless you're implying there's a way that users re-compile+repackage the Iceberg AWS bundle - and that there's a supported way for them to drop dependencies while doing this process (which in my opinion as an active OSS contributor/maintainer would be bizarre).

Reply:
All the AWS dependencies are not packaged together with the iceberg-aws module: https://github.com/apache/iceberg/blob/main/build.gradle#L467-L476. This is because the users typically manage their own AWS SDK version that is very likely different from the one specified by this project.

Reply:
I don't think we're on the same page about this - the S3 Access Grants plugin code is separate from the remaining AWS SDK and therefore can be treated as any other 3rd-party dependency. But this is a moot point with the new change I've introduced.

The size increase after these new changes are just 2.2KB (likely due to the source code introduced itself at this point). Please let me know if these changes are sufficient or if there are any other changes that need to be made. Thanks!

Reply:
> the S3 Access Grants plugin code is separate from the remaining AWS SDK and therefore can be treated as any other 3rd-party dependency. But this is a moot point with the new change I've introduced.

I think it would be quite awkward to not make AWS SDK and AWS SDK plugin consistent. We can add the dependency as a `implementation` as part of the `aws-bundle` package, but this dependency should remain `compileOnly` in the codebase, and make sure that users who don't have the dependency can still use the iceberg-aws library, or any runtime jars that bundles the library like the Spark and Flink runtime jars.